kubernetes/kubernetes.go

```diff
@@ -7,23 +7,82 @@ import (
 	v13 "k8s.io/api/core/v1"
 	v1beta12 "k8s.io/api/extensions/v1beta1"
 	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
 	v12 "k8s.io/client-go/kubernetes/typed/core/v1"
 	"k8s.io/client-go/kubernetes/typed/extensions/v1beta1"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
 	"math"
 	"net/url"
 	"strconv"
-
-	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
+	"sync"
 )
 
+var updateMutex sync.Mutex
+
+func WatchBackends(config *util.Config, updateChan chan util.Update) error {
+	clients, err := GetKubeClient(config.Kubeconfig)
+	if err != nil {
+		return err
+	}
+
+	emptyOptions := v1.ListOptions{}
+
+	ingressWatch, err := clients.ExtensionsV1beta1().Ingresses("").Watch(emptyOptions)
+	if err != nil {
+		return err
+	}
+	defer ingressWatch.Stop()
+
+	serviceWatch, err := clients.CoreV1().Services("").Watch(emptyOptions)
+	if err != nil {
+		return err
+	}
+	defer serviceWatch.Stop()
+
+	endpointWatch, err := clients.CoreV1().Endpoints("").Watch(emptyOptions)
+	if err != nil {
+		return err
+	}
+	defer endpointWatch.Stop()
+
+	var event watch.Event
+	for {
+		select {
+		case event = <-ingressWatch.ResultChan():
+		case event = <-serviceWatch.ResultChan():
+		case event = <-endpointWatch.ResultChan():
+			// ignore endpoint events from kube-system, because:
+			// https://github.com/kubernetes/kubernetes/issues/34627
+			// .. hopefully we don't need ingress on kube-system stuff; if so, we must be more clever here
+			ev, ok := event.Object.(*v13.Endpoints)
+			if ok {
+				if _, ok := config.IgnoreNamespaces[ev.Namespace]; ok {
+					log.Debug("Ignored kubernetes endpoint-API event",
+						zap.String("type", string(event.Type)),
+						zap.String("namespace", ev.Namespace),
+						zap.String("name", ev.Name),
+					)
+					continue
+				}
+			}
+		}
+		log.Debug("Received kubernetes API event",
+			zap.String("type", string(event.Type)),
+			zap.String("object", fmt.Sprint(event.Object)),
+		)
+		updateChan <- util.NewUpdate("api-change")
```

Comment:
would be neat to get the `event.Type` here, I think

Reply:
IMHO. Unless we are using it somewhere, we don't need it on updateChan.

I thought of it too at some point. Until I realized that updateChan is generic and it seems odd to propagate kubernetes-specific info there. Of course you could stretch the argument and say `Type` is sounds generic enough, but then the Update log entry will have fields: `event`, `reason` and `type`. And we sort-of have to find out what to put in each of these fields for all update cases and try to keep it making sense.

Reply:
Oh, sorry, I should have been clearer -- I just meant adding the string-representation to the existing string, as in `util.NewUpdate("api-change-${toString event.Type}")` (as proper go instead of nix, of course :p).

Reply:
eeh.. ok.. I can do that.. not sure if I see the value of polluting the reason-field though :P Now you loose the ability to query: `where reason == 'api-change'` - of course if you only grep or glob-match anyway, it's no problem.

Should I `toLower` it too ? ;) because... events are like either: `ADD`, `DELETE`, `MODIFY` in caps; iirc.

Reply:
Meh, you're probably right. It would need way more information to help with debugging than just that, so it probably doesn't add much (if any) value on its own.

Reply:
Right. :) I was just about to clarify my comment: 
> I think I fail to see the value of `event.Type` especially because you won't be able to see what have changed anyway, unless you log out more information about the object changed.